www/js/lib/zstdec_wrapper.js

```diff
@@ -0,0 +1,284 @@
+        this._inStreamChunkedPos = 0;
+        this._outStreamPos = 0;
+        this._outDataBuf = new Int8Array(new ArrayBuffer(length));
+        this._outDataBufPos = 0;
+        
+        // Initialize inBuffer
+        this._inBuffer = {
+            ptr: null,      /* pointer to this inBuffer structure in w/asm memory */
+            src: null,      /* void* src   < start of input buffer */
+            size: length,   /* size_t size < size of input buffer */
+            pos: 0          /* size_t pos; < position where reading stopped. Will be updated. Necessarily 0 <= pos <= size */
+        };
+        // Reserve w/asm memory for the outBuffer structure
+        this._inBuffer.ptr = mallocOrDie(3 << 2); // 3 x 32bit bytes
+        // DEV: Size of outBuffer is currently set as recommended by zd._ZSTD_DStreamOutSize() below; if you are running into
+        // memory issues, it may be possible to reduce memory consumption by setting asmaller outBuffer size here and
+        // reompiling zstdec.js with lower TOTAL_MEMORY (or just search for INITIAL_MEMORY in zstdec.js and change it)
+        var recOutbufSize = zd._ZSTD_DStreamOutSize();
+        // Initialize outBuffer
+        this._outBuffer = {
+            ptr: null,           /* pointer to this outBuffer structure in asm/wasm memory */
+            dst: null,           /* void* dst   < start of output buffer (pointer) */
+            size: recOutbufSize, /* size_t size < size of output buffer */
+            pos: 0               /* size_t pos  < position where writing stopped. Will be updated. Necessarily 0 <= pos <= size */
+        };
+        this._outBuffer.ptr = mallocOrDie(3 << 2); // 3 x 32bit bytes
+        var ret = zd._ZSTD_initDStream(zd._decHandle);
+        if (zd._ZSTD_isError(ret)) {
+            return Q.reject('Failed to initialize ZSTD decompression');
+        }
+
+        var that = this;
+        return this._readLoop(offset, length).then(function(data) {
+            // DEV: These structures are a known fixed length and could be assigned once, avoiding the need to free them
+            // currently they are re-assigned on each blob request; consider changing this if memory usage appears to grow over time
+            zd._free(that._inBuffer.src);
+            zd._free(that._inBuffer.ptr);
+            zd._free(that._outBuffer.dst);
+            zd._free(that._outBuffer.ptr);
+            // DEV: Freeing zd._decHandle is not needed, and actually increases memory consumption (crashing zstddeclib)
+            // The library explicitly encourages re-using assigned structures and handles
+            // zd._ZSTD_freeDStream(zd._decHandle);
+            busy = false;
+            console.log("Freed all data structures.");
+            return data;
+        });
+    };
+
+    /**
+     * Reads stream of data from file offset for length of bytes to send to the decompresor
+     * This function ensures that only one decompression runs at a time
```

Comment:
It would be clearer to specify that the file offset is the beginning of a compressed cluster...